runtime/vm/classloadersearch.c

```diff
@@ -245,77 +271,26 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
 jclass
 getJimModules(J9VMThread *currentThread)
 {
-	J9JavaVM *vm = NULL;
-	vm = currentThread->javaVM;
-	if (NULL == vm->jimModules) {
-		const J9InternalVMFunctions *vmFuncs = NULL;
-		/* no need for mutual exclusion since repeating this is harmless */
-		J9Class *j9JimModules = NULL;
-		internalEnterVMFromJNI(currentThread);
-		vmFuncs = vm->internalVMFunctions;
-		j9JimModules = vmFuncs->internalFindKnownClass(currentThread,
-				J9VMCONSTANTPOOL_JDKINTERNALMODULEMODULES,
-				J9_FINDKNOWNCLASS_FLAG_INITIALIZE | J9_FINDKNOWNCLASS_FLAG_NON_FATAL);
-		if (NULL != j9JimModules) {
-			/*
-			 * The address of j9class->classObject is considered to be a JNI global reference.
-			 * Note that the class in question is not a candidate for unloading.
-			 */
-			vm->jimModules = (jclass)&j9JimModules->classObject;
+	J9JavaVM *vm = currentThread->javaVM;
+	jclass jimModules = vm->jimModules;
+	if (NULL == jimModules) {
+		JNIEnv *env = (JNIEnv*)currentThread;
+		jclass modulesClass = (*env)->FindClass(env, "jdk/internal/module/Modules");
+		if (NULL == modulesClass) {
+			(*env)->ExceptionClear(env);
 		} else {
-			currentThread->currentException = NULL;
-			currentThread->privateFlags &= ~J9_PRIVATE_FLAGS_REPORT_EXCEPTION_THROW;
+			/* no need for mutual exclusion since repeating this is harmless */
+			J9Class *j9JimModules = NULL;
+			internalEnterVMFromJNI(currentThread);
+			j9JimModules = J9VM_J9CLASS_FROM_JCLASS(currentThread, modulesClass);
+			/* The address of j9class->classObject is considered to be a JNI global reference,
+			 * but it won't keep the class alive.  That's OK in this case since the class in
+			 * question is not a candidate for unloading.
+			 */
+			vm->jimModules = jimModules = (jclass)&j9JimModules->classObject;
```

Comment:
This is iffy as it assumes the J9Class will never be freed (today's behaviour).  We've been talking about freeing the obsolete classes and this cause a hard to find bug.

Reply:
Ah, this is just refactored existing code... something to be cleaned up, not necessarily as part of this PR.  If you're not going to fix this now, let me know and I'll open an issue for it.

Reply:
It's iffy for a class that we don't know everything about - in this case, it's a boot class (or whatever the JDK9 equivalent is), so we are sure it's not a candidate for unloading.

Reply:
Note, in JDK9+ not every `java.*` class is loaded by the bootloader.  Some have been moved to the app loader.  (This one is loaded by the bootloader)

My concern is more with future bugs when we allow redefined J9Classes to be GC'd.... though this may not be a concern with FastHCR due to keeping the J9Class stable.

All that to say, we should at minimum open a issue to remember to look at this again,

Reply:
So, I don't want an issue for it - this was a change we made to avoid the need to lock and allocate a global ref.

Reply:
Why are we trying to avoid creating the global ref?  Is there a usecase that demonstrates the current solution is better?  I would expect the extra cost of the global ref to be in the noise.

Reply:
It was to avoid the need for a single-use monitor.  Since you're clearly upset by this, I'll change it back (though I'll reuse an existing monitor).